File: backend-node/src/services/audit.ts
import { query } from '../db.ts'

export async function logAction(opts: {
  entityType: string
  entityId?: string | null
  action: string
  userId?: string | null
  invoiceId?: string | null
  details?: Record<string, unknown> | null
}): Promise<void> {
  await query(
    `INSERT INTO audit_logs (entity_type, entity_id, action, user_id, invoice_id, details)
     VALUES ($1, $2, $3, $4, $5, $6)`,
    [
      opts.entityType,
      opts.entityId ?? null,
      opts.action,
      opts.userId ?? null,
      opts.invoiceId ?? null,
      opts.details ? JSON.stringify(opts.details) : null,
    ]
  )
}

File: backend/app/models/audit.py
import uuid

from sqlalchemy import Column, ForeignKey, JSON, String
from sqlalchemy.dialects.postgresql import UUID
from sqlalchemy.orm import relationship

from .base import Base, TimestampMixin


class AuditLog(Base, TimestampMixin):
    __tablename__ = "audit_logs"

    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    entity_type = Column(String, nullable=False)
    entity_id = Column(UUID(as_uuid=True), nullable=True)
    action = Column(String, nullable=False)
    user_id = Column(UUID(as_uuid=True), ForeignKey("users.id"), nullable=True)
    invoice_id = Column(UUID(as_uuid=True), ForeignKey("invoices.id"), nullable=True)
    details = Column(JSON, nullable=True)

    user = relationship("User", back_populates="audit_logs")
    invoice = relationship("Invoice", back_populates="audit_logs")
